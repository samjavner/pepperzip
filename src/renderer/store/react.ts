import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
