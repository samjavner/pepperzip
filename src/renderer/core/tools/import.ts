import { remote } from "electron";
import * as uuid from "uuid";
import * as fdx from "../../formats/fdx";
import * as mmf from "../../formats/mmf/parser";
import * as mx2 from "../../formats/mx2";
import * as mxp from "../../formats/mxp/parser";
import * as paprika from "../../formats/paprika/parser";
import * as schemaorg from "../../formats/schema.org";
import { Dispatch, Store, UseStore } from "../../store";
import { RecipeBox } from "../recipes/recipeBox";

export type Import = Store<Model, Update, Commands>;

export function useImport(
    useStore: UseStore<Model, Update, Commands>,
    recipeBox: RecipeBox
): Import {
    return useStore({
        init,
        update,
        createCommands: createCommands(recipeBox),
        memo: [recipeBox],
    });
}

// MODEL

export interface Model {
    isImporting: boolean;
}

export const init: Model = {
    isImporting: false,
};

// UPDATE

type Update = typeof update;

export const update = {
    importRequest(model: Model): Model {
        return {
            ...model,
            isImporting: true,
        };
    },
    importSuccess(model: Model): Model {
        return {
            ...model,
            isImporting: false,
        };
    },
};

// COMMANDS

type Commands = ReturnType<ReturnType<typeof createCommands>>;

export const createCommands = (recipeBox: RecipeBox) => (
    model: Model,
    dispatch: Dispatch<Update>
) => {
    async function importRecipes() {
        const paths = remote.dialog.showOpenDialog(remote.getCurrentWindow(), {
            filters: [
                {
                    name: "Recipe files",
                    extensions: [
                        "fdx",
                        "json",
                        "mmf",
                        "mx2",
                        "mxp",
                        "paprikarecipes",
                    ],
                },
            ],
        });

        if (!paths || paths.length === 0) {
            return;
        }

        await dispatch.importRequest();
        const path = paths[0];

        if (path.toLowerCase().endsWith(".mmf")) {
            const recipes = mmf.parseFileWithSource(path);
            await recipeBox.addMultiple(
                recipes.map(recipe => ({
                    id: uuid.v4(),
                    name: recipe[1].title || "Imported Recipe",
                    ingredients: recipe[1].ingredients
                        .map(
                            ingredient =>
                                ingredient.type === "ingredient"
                                    ? (
                                          ingredient.quantity +
                                          " " +
                                          (
                                              ingredient.unit +
                                              " " +
                                              ingredient.text
                                          ).trim()
                                      ).trim()
                                    : ingredient.text
                        )
                        .join("\n"),
                    directions: recipe[1].directions.join("\n"),
                }))
            );
        } else if (path.toLowerCase().endsWith(".mxp")) {
            const recipes = mxp.parseFileWithSource(path);
            // tslint:disable-next-line:no-console
            console.log(recipes);
        } else if (path.toLowerCase().endsWith(".mx2")) {
            const recipes = await mx2.parseFile(path);
            // tslint:disable-next-line:no-console
            console.log(recipes);
        } else if (path.toLowerCase().endsWith(".fdx")) {
            const recipes = await fdx.parseFile(path);
            // tslint:disable-next-line:no-console
            console.log(recipes);
        } else if (path.toLowerCase().endsWith(".paprikarecipes")) {
            const recipes = paprika.parseFile(path);
            // tslint:disable-next-line:no-console
            console.log(recipes);
        } else if (path.toLowerCase().endsWith(".json")) {
            const recipes = schemaorg.parseFile(path);
            // tslint:disable-next-line:no-console
            console.log(recipes);
        }

        await dispatch.importSuccess();
    }

    return {
        import: importRecipes,
    };
};
