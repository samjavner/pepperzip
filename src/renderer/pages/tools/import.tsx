import { remote } from "electron";
import * as React from "react";
import * as fdx from "../../formats/fdx";
import * as mmf from "../../formats/mmf/parser";
import * as mx2 from "../../formats/mx2";
import * as mxp from "../../formats/mxp/parser";
import * as paprika from "../../formats/paprika/parser";
import * as schemaorg from "../../formats/schema.org";

// VIEW

export const Page: React.FunctionComponent = () => (
    <div>
        <button
            className="button"
            onClick={async () => {
                const paths = remote.dialog.showOpenDialog(
                    remote.getCurrentWindow(),
                    {
                        filters: [
                            {
                                name: "Recipe files",
                                extensions: [
                                    "fdx",
                                    "json",
                                    "mmf",
                                    "mx2",
                                    "mxp",
                                    "paprikarecipes",
                                ],
                            },
                        ],
                    }
                );

                if (!paths || paths.length === 0) {
                    return;
                }

                const path = paths[0];

                if (path.toLowerCase().endsWith(".mmf")) {
                    const recipes = mmf.parseFileWithSource(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                } else if (path.toLowerCase().endsWith(".mxp")) {
                    const recipes = mxp.parseFileWithSource(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                } else if (path.toLowerCase().endsWith(".mx2")) {
                    const recipes = await mx2.parseFile(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                } else if (path.toLowerCase().endsWith(".fdx")) {
                    const recipes = await fdx.parseFile(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                } else if (path.toLowerCase().endsWith(".paprikarecipes")) {
                    const recipes = paprika.parseFile(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                } else if (path.toLowerCase().endsWith(".json")) {
                    const recipes = schemaorg.parseFile(path);
                    // tslint:disable-next-line:no-console
                    console.log(recipes);
                }
            }}
        >
            Import
        </button>
    </div>
);
