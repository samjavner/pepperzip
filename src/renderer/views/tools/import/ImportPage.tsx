import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../../app/actions";
import { State } from "../../../app/state";

function mapStateToProps(state: State) {
    return {};
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {};
}

type ImportPageProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const ImportPage: React.SFC<ImportPageProps> = props => (
    <div>Import</div>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ImportPage);
