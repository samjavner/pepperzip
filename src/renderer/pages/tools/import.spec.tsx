import * as React from "react";
import * as renderer from "react-test-renderer";
import * as toolsImport from "./import";

describe("import", () => {
    describe("Page", () => {
        it("renders a button with the text 'Import'", () => {
            const component = renderer.create(<toolsImport.Page />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
