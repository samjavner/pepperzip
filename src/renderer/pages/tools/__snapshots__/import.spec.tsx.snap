// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import Page renders a button with the text 'Import' 1`] = `
<div>
  <button
    className="button"
    onClick={[Function]}
  >
    Import
  </button>
</div>
`;
